### PR TITLE
Backport BEAM-985 and BEAM-997: Support BigQuery DATE, DATETIME and TIME types

### PR DESCRIPTION
**Pull request by peihe**
Backport:
https://github.com/apache/incubator-beam/pull/997
https://github.com/apache/incubator-beam/pull/985

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @lukecwik

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/util/AvroUtils.java [hunk 1]
```diff
@@ -271,11 +272,15 @@ private static Object convertRequiredField(
             .put("BOOLEAN", Type.BOOLEAN)
             .put("TIMESTAMP", Type.LONG)
             .put("RECORD", Type.RECORD)
-            .build();
+            .put("DATE", Type.STRING)
+            .put("DATETIME", Type.STRING)
+            .put("TIME", Type.STRING)
+        .build();
```
Comment (lukecwik) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/AvroUtils.java]: spacing

Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/util/AvroUtils.java]: done

**Comment by lukecwik** [comment]
LGTM, will merge after Travis run

**Comment by lukecwik** [comment]
All the travis runs are failing with:
java.lang.ClassNotFoundException: com.google.cloud.dataflow.examples.WordCount
	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:281)
	at java.lang.Thread.run(Thread.java:744)

**Comment by dhalperi** [comment]
The wordcount message is usually indicative of an error earlier in the job. In this case:

```
ailed tests: 
  AvroUtilsTest.testReadSchemaString:110 expected:<9> but was:<12>
Tests in error: 
com.google.cloud.dataflow.sdk.util.BigQueryTableRowIteratorTest.testReadFromQueryNoTables(com.google.cloud.dataflow.sdk.util.BigQueryTableRowIteratorTest)
  Run 1: BigQueryTableRowIteratorTest.testReadFromQueryNoTables:275 » IllegalState Expe...
  Run 2: BigQueryTableRowIteratorTest.tearDown:108 
No interactions wanted here:
-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIteratorTest.tearDown(BigQueryTableRowIteratorTest.java:108)
But found this interaction on mock 'mockClient':
-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.executeQueryAndWaitForCompletion(BigQueryTableRowIterator.java:410)
***
For your reference, here is the list of all invocations ([?] - means unverified).
1. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.executeQueryAndWaitForCompletion(BigQueryTableRowIterator.java:410)
2. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.createDataset(BigQueryTableRowIterator.java:368)
3. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.executeQueryAndWaitForCompletion(BigQueryTableRowIterator.java:446)
4. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.executeQueryAndWaitForCompletion(BigQueryTableRowIterator.java:452)
5. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.getTable(BigQueryTableRowIterator.java:347)
6. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.advance(BigQueryTableRowIterator.java:174)
7. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.deleteTable(BigQueryTableRowIterator.java:378)
8. [?]-> at com.google.cloud.dataflow.sdk.util.BigQueryTableRowIterator.deleteDataset(BigQueryTableRowIterator.java:388)
```

**Comment by lukecwik** [comment]
I see, it seems like our travis wordcount script should exit instead of proceeding after failure

**Pull request merged**